target/linux/lantiq/patches-5.10/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch

```diff
@@ -0,0 +1,125 @@
+From d39d31486432bb3e54e4cb83744d99c3f5d773c6 Mon Sep 17 00:00:00 2001
+From: Thomas Nixon <tom@tomn.co.uk>
+Date: Sat, 3 Jul 2021 15:57:03 +0100
+Subject: [PATCH] net: lantiq: enable baby jumbo frames on xrx200
+
+xrx200 max MTU is reduced so that it works correctly when set to the
+max, and the max MTU of the switch is increased.
+
+The switch driver now enables non-standard MTUs on a per-port basis,
+with the overall frame size set based on the cpu port.
+
+When the MTU is not changed, this should have no effect.
+
+Signed-off-by: Thomas Nixon <tom@tomn.co.uk>
+---
+ drivers/net/dsa/lantiq_gswip.c       | 42 ++++++++++++++++++++++++++--
+ drivers/net/ethernet/lantiq_xrx200.c |  4 ++-
+ 2 files changed, 42 insertions(+), 4 deletions(-)
+
+diff --git a/drivers/net/dsa/lantiq_gswip.c b/drivers/net/dsa/lantiq_gswip.c
+index 93c7fa1fd4cb..1e67562153f9 100644
+--- a/drivers/net/dsa/lantiq_gswip.c
++++ b/drivers/net/dsa/lantiq_gswip.c
+@@ -238,6 +238,8 @@
+ 
+ #define XRX200_GPHY_FW_ALIGN	(16 * 1024)
+ 
++#define GSWIP_MAX_PACKET_LENGTH	9600
++
+ struct gswip_hw_info {
+ 	int max_ports;
+ 	int cpu_port;
+@@ -851,9 +853,6 @@ static int gswip_setup(struct dsa_switch *ds)
+ 	gswip_switch_mask(priv, 0, GSWIP_PCE_PCTRL_0_INGRESS,
+ 			  GSWIP_PCE_PCTRL_0p(cpu_port));
+ 
+-	gswip_switch_mask(priv, 0, GSWIP_MAC_CTRL_2_MLEN,
+-			  GSWIP_MAC_CTRL_2p(cpu_port));
+-	gswip_switch_w(priv, VLAN_ETH_FRAME_LEN + 8, GSWIP_MAC_FLEN);
+ 	gswip_switch_mask(priv, 0, GSWIP_BM_QUEUE_GCTRL_GL_MOD,
+ 			  GSWIP_BM_QUEUE_GCTRL);
+ 
+@@ -870,6 +869,8 @@ static int gswip_setup(struct dsa_switch *ds)
+ 		return err;
+ 	}
+ 
++	ds->mtu_enforcement_ingress = true;
++
+ 	gswip_port_enable(ds, cpu_port, NULL);
+ 	return 0;
+ }
+@@ -1426,6 +1427,39 @@ static int gswip_port_fdb_dump(struct dsa_switch *ds, int port,
+ 	return 0;
+ }
+ 
++static int gswip_port_max_mtu(struct dsa_switch *ds, int port)
++{
++	/* includes 8 bytes for special header */
++	return GSWIP_MAX_PACKET_LENGTH - VLAN_ETH_HLEN - ETH_FCS_LEN;
++}
++
++static int gswip_port_change_mtu(struct dsa_switch *ds, int port, int new_mtu)
++{
++	struct gswip_priv *priv = ds->priv;
++	int cpu_port = priv->hw_info->cpu_port;
++
++	/* cpu port always has maximum mtu of user ports, so use it to set
++	 * switch frame size, including 8 byte special header
++	 */
++	if (port == cpu_port) {
++		new_mtu += 8;
++		gswip_switch_w(priv, VLAN_ETH_HLEN + new_mtu + ETH_FCS_LEN,
++			       GSWIP_MAC_FLEN);
```

Comment:
This is not configuring the MTU for the CPU port, but what the switch will handle internally.

Reply:
The comment above aims to address this:

> cpu port always has maximum mtu of user ports, so use it to set switch frame size, including 8 byte special header

Does this make sense? Doing it this way, rather than finding the maximum port MTU was suggested by  LGA1150 above.

Reply:
I think GSWIP is slightly different than MT7530 (which is what @LGA1150 worked on).
My understanding is that:
- both have a global register for the maximum MTU which is then valid across all ports. on GSWIP the register is GSWIP_MAC_FLEN
- in addition to that GSWIP can limit the frame length per port to 1518 + VLAN tag length (which is not possible on MT7530 from what I understand). on GSWIP the register is GSWIP_MAC_CTRL_2p(port)

Unfortunately I am not an expert on this topic. when in doubt: the netdev maintainers are very helpful and respond fast (I always get a reply within 24 hours)